codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/customizations/SigV4EventStreamDecorator.kt

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.rust.codegen.server.smithy.customizations
+
+import software.amazon.smithy.aws.traits.auth.SigV4Trait
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.model.knowledge.ServiceIndex
+import software.amazon.smithy.model.shapes.MemberShape
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.shapes.Shape
+import software.amazon.smithy.model.shapes.ShapeId
+import software.amazon.smithy.rust.codegen.core.rustlang.RustType
+import software.amazon.smithy.rust.codegen.core.rustlang.Writable
+import software.amazon.smithy.rust.codegen.core.rustlang.rustTemplate
+import software.amazon.smithy.rust.codegen.core.rustlang.writable
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeConfig
+import software.amazon.smithy.rust.codegen.core.smithy.RustSymbolProvider
+import software.amazon.smithy.rust.codegen.core.smithy.WrappingSymbolProvider
+import software.amazon.smithy.rust.codegen.core.smithy.generators.http.HttpBindingCustomization
+import software.amazon.smithy.rust.codegen.core.smithy.generators.http.HttpBindingSection
+import software.amazon.smithy.rust.codegen.core.smithy.mapRustType
+import software.amazon.smithy.rust.codegen.core.smithy.rustType
+import software.amazon.smithy.rust.codegen.core.smithy.traits.SyntheticInputTrait
+import software.amazon.smithy.rust.codegen.core.util.getTrait
+import software.amazon.smithy.rust.codegen.core.util.isEventStream
+import software.amazon.smithy.rust.codegen.core.util.isInputEventStream
+import software.amazon.smithy.rust.codegen.server.smithy.customize.ServerCodegenDecorator
+
+/**
+ * Decorator that adds SigV4 event stream unsigning support to server code generation.
+ */
+class SigV4EventStreamDecorator : ServerCodegenDecorator {
+    override val name: String = "SigV4EventStreamDecorator"
+    override val order: Byte = 0
+
+    override fun httpCustomizations(
+        symbolProvider: RustSymbolProvider,
+        protocol: ShapeId,
+    ): List<HttpBindingCustomization> {
+        return listOf(SigV4EventStreamCustomization(symbolProvider))
+    }
+
+    override fun symbolProvider(base: RustSymbolProvider): RustSymbolProvider {
+        // We need access to the service shape to check for SigV4 trait, but the base interface doesn't provide it.
```

Comment:
Nit: We can use the function `transformModel` to detect whether the service uses sigV4 authentication or not.

Reply:
Considered, decided it was too complex